### PR TITLE
Updates unit tests and linting

### PR DESCRIPTION
**Pull request by jasonb5**
- Split up `scripts_regression_tests.py` into smaller units under
  `scripts/lib/CIME/tests`
- Update all unit tests to inherit common base class `BaseTestCase`
- Retains `scripts_regression_tests.py` under `scripts/lib/CIME/tests`,
  maintains same CLI
- Adds `conftest.py` to provide same CLI as `scripts_regression_tests.py`
  for `pytest`
- Adds configuration for `coverage` via `pytest-cov`
- Updates github actions to use `pytest` and report coverage
- Adds `pre-commit` to handle `pylint` and ran against PR/merges into
  master branch

### Running tests
```
python scripts/lib/CIME/tests/scripts_regression_tests.py
# or
python -m pip install pytest pytest-cov
pytest -vvv
```
### Linting
```
python -m pip install pre-commit
pre-commit run -a
```

Test suite: 
Test baseline: n/a
Test namelist changes: n/a
Test status: bfb

Fixes [CIME Github issue #] n/a
User interface changes?: n/a
Update gh-pages html (Y/N)?: Y

### CODE REVIEW

**Comment by jedwards4b** [comment]
When I try to run the standalone testing using:
  CIME_MODEL=cesm python scripts/lib/CIME/tests/scripts_regression_tests.py 
I get:
  Machine object has no machine defined
  ERROR: No list found for MPILIBS on machine cheyenne with attributes None
When I try the same syntax with create_test as in CIME_MODEL=cesm ./create_test SMS.f19_g17.X
MPILIBS is found and it seems to work as expected.

**Comment by jedwards4b** [comment]
Note that I edited machines.py to get the exact ERROR message above.  The original was
ERROR: No list found for MPILIBS on machine cheyenne

**Comment by jasonb5** [comment]
@jedwards4b I'm looking into this error.

**Comment by jasonb5** [comment]
@jedwards4b Could you try this again, it should be fixed. I was setting a default TEST_COMPILER and TEST_MPILIB when they were undefined on the cli. This would bypass the original logic to determine those when undefined.

**Comment by jedwards4b** [comment]
@jasonb5 Thanks, this is an improvement.  Another issue - when running the current scripts_regression_tests.py there is output at the beginning of the run:
```
Testing commit b0f7080a9597d4d5ecf46aa7a4b431670835a3ea
Using cime_model = cesm
Testing machine = cheyenne
Test root: /glade/scratch/jedwards/scripts_regression_test.20211130_113942
Test driver: nuopc
Python version 3.7.9 (default, Dec 20 2020, 13:26:20) 
[GCC 9.1.0]
```

this is very useful information and I don't see an analog with the new method.   In fact I just started a test with CIME_COMPILER=gnu, but I can't tell if it's actually using the gnu compilers.

**Comment by jasonb5** [comment]
@jedwards4b Fixed the provenance and it's available when using `pytest` as well.

**Review by jgfouca — APPROVED** [review]
Nice job!

**Comment by jgfouca** [comment]
@jasonb5 , does this:

```
python -m pip install pre-commit
pre-commit run -a
```

... replace our code-checker tool?

**Comment by jasonb5** [comment]
@jgfouca Yes this replaces the code-checker tool. It runs `pylint` with the same configuration as the code-checker.

File: scripts/lib/CIME/tests/scripts_regression_tests.py [hunk 1]
```diff
@@ -0,0 +1,226 @@
+#!/usr/bin/env python3
+
+"""
+Script containing CIME python regression test suite. This suite should be run
+to confirm overall CIME correctness.
+"""
+
+import glob, os, re, shutil, signal, sys, tempfile, \
```
Comment (billsacks) [inline scripts/lib/CIME/tests/scripts_regression_tests.py:8]: - [ ] A lot of the imports in this file are no longer referenced.

**Review by billsacks — CHANGES_REQUESTED** [review]
Thank you very much for these changes! In addition to my two inline comments, I have two related overall comments / requests. It may be that taking care of one of these makes the other irrelevant:

- [ ] I appreciate the splitting up of scripts_regression_tests. However, I would find it helpful if we distinguished between two classes of tests in `scripts/lib/CIME/tests`: unit tests vs. more system-y tests. The former would be tests that run nearly instantaneously whereas the latter would be tests that take considerable time to run (e.g., order seconds or more). For the most part, I imagine that the latter would include tests that actually set up a real case. In other projects I have worked on, we make this distinction with a prefix in the file name – so `test_unit_*.py` or `test_sys_*.py`. This lends itself to running just the fast unit tests via python's unittest discover mechanism. Alternatively, there could be subdirectories for these two classes of tests.

- [ ] Maybe connected with the above comment: it used to work to run `python3 -m unittest discover -s CIME/tests -t .`. This now leads to a lot of test failures, presumably because of tests that need to be run via `scripts_regression_tests.py` (using its `configure_tests` function). So another distinction that could be made is between tests that can be run directly vs. those that can only be run via `scripts_regression_tests`. This distinction might make the above suggestion (for a distinction between unit vs. sys tests) unnecessary. Again, the distinction could be made via a file name pattern or subdirectories.

File: scripts/lib/CIME/tests/scripts_regression_tests.py (hunk 1 above)
Comment (jasonb5) [reply scripts/lib/CIME/tests/scripts_regression_tests.py:8]: The plan is resolve this in my next PR that'll be targeting code formatting/style enforcement. I didn't want to add extra changes to this PR and rather tackle it all in a separate one.

File: doc/source/users_guide/testing.rst [hunk 2]
```diff
@@ -455,26 +455,37 @@ CIME's scripts regression tests
...
 
 Run full suite::
 
-  ./scripts_regression_tests.py
+  python scripts/lib/CIME/tests/scripts_regression_tests.py
 
 Run a test class::
 
-  ./scripts_regression_tests.py K_TestCimeCase
+  python scripts/lib/CIME/tests/scripts_regression_tests.py K_TestCimeCase
```
[the remark this answers is not in the document]
Comment (jasonb5) [reply doc/source/users_guide/testing.rst:468]: Good catch this doesn't work anymore, I'll add this feature. It'll probably be along the lines of `python scripts/lib/CIME/tests/scripts_regression_tests.py CIME.tests.test_cime_case`.

**Comment by jasonb5** [comment]
@billsacks Adding prefixes to the tests is a great idea and would solve `python -m unittest` not working. Indeed there are now tests in the folder that require some setup that is provided by `pytest` and `scripts_regression_tests.py`, these are mostly e2e (sys) tests.

I'll go through the tests and add these prefixes.

File: scripts/lib/CIME/tests/scripts_regression_tests.py (hunk 1 above)
Comment (billsacks) [reply scripts/lib/CIME/tests/scripts_regression_tests.py:8]: Sounds good - thanks.

File: doc/source/users_guide/testing.rst (hunk 2 above)
[the remark this answers is not in the document]
Comment (billsacks) [reply doc/source/users_guide/testing.rst:468]: Sounds great - thanks. I personally am fine with whatever syntax is most convenient.

[the remark this answers is not in the document]
Comment (jasonb5) [reply doc/source/users_guide/testing.rst]: Updated the docs with the correct syntax.

**Review by billsacks — CHANGES_REQUESTED** [review]
This is great @jasonb5 ! Thanks a lot for making the recent changes.

My only remaining request is related to the documentation: This doesn't quite work:

```
pytest -vvv scripts/lib/CIME/tests/test_unit*
```

because there are some tests in subdirectories. I'm not sure if this directory organization makes sense, but my original thinking was that the files under `scripts/lib/CIME/tests` could be in a directory structure that mirrors that of `scripts/lib/CIME`. So there are some files in `scripts/lib/CIME/tests/SystemTests` that test files in `scripts/lib/CIME/SystemTests` (and even one level deeper than that), and similarly for `XML`.

The documented `pytest` command skips these. Is there a way to make it find files recursively?

I *was* able to successfully run this command (from the `scripts/lib` directory):

```
python3 -m unittest discover -s CIME/tests -t . -p 'test_unit*'
```

but it seems like you are encouraging the use of pytest, and I don't know the equivalent command with pytest.

**Comment by jasonb5** [comment]
@billsacks I'm inclined to just remove the directories under `tests` and make it a flat directory. The two prefixes provide good organization between `unit` and `sys` tests. If finer grain filtering is required the individual test suite/test case can be specified under each tool.

**Comment by billsacks** [comment]
I'm fine with that.

**Review by billsacks — APPROVED** [review]
This looks good to me now. Thanks a lot for all of your work on it @jasonb5 !

**Review by jedwards4b — APPROVED** [review]

**Pull request merged**